src/components/ha-form/ha-form-optional_actions.ts

```diff
@@ -0,0 +1,162 @@
+
+  @property({ attribute: false }) public hass!: HomeAssistant;
+
+  @property({ attribute: false }) public data!: HaFormDataContainer;
+
+  @property({ attribute: false }) public schema!: HaFormOptionalActionsSchema;
+
+  @property({ type: Boolean }) public disabled = false;
+
+  @property({ attribute: false }) public computeLabel?: (
+    schema: HaFormSchema,
+    data?: HaFormDataContainer
+  ) => string;
+
+  @property({ attribute: false }) public computeHelper?: (
+    schema: HaFormSchema
+  ) => string;
+
+  @property({ attribute: false }) public localizeValue?: (
+    key: string
+  ) => string;
+
+  @state() private _displayActions?: string[];
+
+  public async focus() {
+    await this.updateComplete;
+    this.renderRoot.querySelector("ha-form")?.focus();
+  }
+
+  protected updated(changedProps: PropertyValues): void {
+    super.updated(changedProps);
+    if (changedProps.has("data")) {
+      const displayActions = this._displayActions ?? NO_ACTIONS;
+      const hiddenActions = this._hiddenActions(
+        this.schema.schema,
+        displayActions
+      );
+      this._displayActions = [
+        ...displayActions,
+        ...hiddenActions.filter((name) => name in this.data),
+      ];
+    }
+  }
+
+  private _hiddenActions = memoizeOne(
+    (schema: readonly HaFormSchema[], displayActions: string[]): string[] =>
+      schema
+        .map((item) => item.name)
+        .filter((name) => !displayActions.includes(name))
+  );
+
+  private _displaySchema = memoizeOne(
+    (
+      schema: readonly HaFormSchema[],
+      displayActions: string[]
+    ): HaFormSchema[] =>
+      schema.filter((item) => displayActions.includes(item.name))
+  );
+
+  public render(): TemplateResult {
+    const displayActions = this._displayActions ?? NO_ACTIONS;
+
+    const schema = this._displaySchema(
+      this.schema.schema,
+      this._displayActions ?? []
+    );
+
+    const hiddenActions = this._hiddenActions(
+      this.schema.schema,
+      displayActions
+    );
+
+    return html`
+      ${schema.length > 0
+        ? html`
+            <ha-form
+              .hass=${this.hass}
+              .data=${this.data}
+              .schema=${schema}
+              .disabled=${this.disabled}
+              .computeLabel=${this.computeLabel}
+              .computeHelper=${this.computeHelper}
+              .localizeValue=${this.localizeValue}
+            ></ha-form>
+          `
+        : nothing}
+      ${hiddenActions.length > 0
+        ? html`
+            <ha-button-menu
+              @action=${this._handleAddAction}
+              fixed
+              @closed=${stopPropagation}
+            >
+              <ha-button slot="trigger">
+                ${this.localize?.("ui.components.form-optional-actions.add") ||
+                "Add interaction"}
+              </ha-button>
+              ${hiddenActions.map((action) => {
+                const actionSchema = this.schema.schema.find(
```

Comment:
Can instead of doing a find every render for every item make a map `{actionName: actionSchema}`?

Reply:
Also, this only seems to be used if `computeLabel` is set, we should skip this if it is not set